src/main/scala/viper/gobra/translator/encodings/StringEncoding.scala

```diff
@@ -0,0 +1,168 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2021 ETH Zurich.
+
+package viper.gobra.translator.encodings
+
+import org.apache.commons.codec.binary.Hex
+import org.bitbucket.inkytonik.kiama.==>
+import viper.gobra.ast.{internal => in}
+import viper.gobra.theory.Addressability.{Exclusive, Shared}
+import viper.gobra.translator.Names
+import viper.gobra.translator.interfaces.{Collector, Context}
+import viper.gobra.translator.util.ViperWriter.CodeLevel.unit
+import viper.gobra.translator.util.ViperWriter.CodeWriter
+import viper.silver.{ast => vpr}
+
+class StringEncoding extends LeafTypeEncoding {
+
+  import viper.gobra.translator.util.TypePatterns._
+
+  private val domainName: String = Names.stringsDomain
+  private var strLengths: Map[String, Int] = Map.empty
+  private var funcs: Map[String, vpr.DomainFunc] = Map.empty
+  private val stringBeginning: String = "stringLit"
+
+  /**
+    * Translates a type into a Viper type.
+    */
+  override def typ(ctx: Context): in.Type ==> vpr.Type = {
+    case ctx.String() / Exclusive => vpr.Int
+    case ctx.String() / Shared => vpr.Ref
+  }
+
+  /**
+    * Encodes expressions as values that do not occupy some identifiable location in memory.
+    *
+    * [ dfltVal: string° ] -> stringLitDefault()
+    * [ strLit: string° ] -> stringLitX() where X is a unique suffix dependant on the value of the string literal
+    * [ len(s: string) ] -> strLen([s])
+    * [ (s1: string) + (s2: string) ] -> strConcat([ s1 ], [ s2 ])
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.expr(ctx)) {
+      case (e: in.DfltVal) :: ctx.String() / Exclusive =>
+        val encodedStr = stringBeginning + "Default"
```

Comment:
Which string is actually the default? Then this should be used instead of "Default". Since your makeFunc results in unique functions, you get an unsoundness otherwiese, e.g. `var x string; assert x == "Default"` would verify.

Reply:
I do not think that this is unsound. The reason is that, for every string other than the default, the suffix of the unique function is the hexadecimal representation of the string. This serves the purpose of having injective suffixes with valid characters (if I just added the original string value as a suffix, it could contain spaces or chars that are not acceptable in identifiers). Because "Default" is not the hexadecimal representation of any string, there should not be a problem.

Reply:
But then you can show the following: `var s string; assert s != ""`

Reply:
I realized that there is a issue here though. Given the current way we encode things, the string literal "" and DfltVal will lead to different encodings even though they are meant to be the same. As such, the following verifies even though it shouldn't:
```go
    a4 := ""
    var a5 string
    assert a4 != a5
```
This has been fixed in the latest commit